TypeScript issues with `moduleResolution: "nodenext"` and `declaration: true`

### Guidelines

- [X] Please search other issues to make sure this bug has not already been reported.

### Describe the bug

When using TypeScript with `"moduleResolution: "nodenext"` and `"declaration": true"`, it fails to compile the example from `README`

### Steps to reproduce

`main.ts`
```ts
import ModernError from 'modern-errors'
export const BaseError = ModernError.subclass('BaseError')
```


`tsconfig.json`
```json
{
  "compilerOptions": {
    "target": "esnext",
    "module": "esnext",
    "moduleResolution": "nodenext",
    "declaration": true
  }
}
```

error:
```
main.ts:3:14 - error TS2742: The inferred type of 'BaseError' cannot be named without a reference to './node_modules/modern-errors/build/src/subclass/create.js'. This is likely not portable. A type annotation is necessary.

3 export const BaseError = ModernError.subclass('BaseError')
               ~~~~~~~~~

main.ts:3:14 - error TS2742: The inferred type of 'BaseError' cannot be named without a reference to './node_modules/modern-errors/build/src/subclass/custom.js'. This is likely not portable. A type annotation is necessary.

3 export const BaseError = ModernError.subclass('BaseError')
               ~~~~~~~~~


Found 2 errors in the same file, starting at: main.ts:3
```

I can make it work in two ways - 
1. Disable `declaration`. However, this means my library cannot publish types and that's not great. 
2. Change to `"moduleResolution": "node"`. However, this means I cannot use the `exports` settings in my dependencies.

With `"moduleResolution":"node"` it generates these types for `main.d.ts`:
```ts
export declare const BaseError: import("modern-errors/build/src/subclass/create").ErrorSubclassCore<[], {}, import("modern-errors/build/src/subclass/custom").CustomClass>;
```

Which hints at the reason for failure: TypeScript has to tap into internal files to reference un-exported types to get `ErrorSubclassCore` (and, in my real project, `CustomClass`). With `"moduleResolution":"node"`, this deep linking is not possible, so the TS compiler gives up.

### Environment

```
  System:
    OS: macOS 13.1
    CPU: (8) arm64 Apple M1
    Memory: 64.61 MB / 8.00 GB
    Shell: 5.8.1 - /bin/zsh
  Binaries:
    Node: 18.14.0 - ~/.asdf/installs/nodejs/18.14.0/bin/node
    Yarn: 1.22.19 - ~/.asdf/shims/yarn
    npm: 9.3.1 - ~/.asdf/plugins/nodejs/shims/npm
```

### Pull request (optional)

- [ ] I can submit a pull request.

--- a/docs/typescript.md
+++ b/docs/typescript.md
@@ -172,3 +172,8 @@ generic parameter, such as `ErrorClass<[typeof plugin]>`.
 They should only be used to type unknown error instances and classes, when no
 variable nor [type inference](#type-inference) is available. For example, they
 can be useful when [creating plugins](plugins.md#typescript).
+
+## Known limitations and issues
+
+For known limitations and issues related to TypeScript's current capabilities,
+please see the comments in [src/main.d.ts](../src/main.d.ts).

--- a/src/main.d.ts
+++ b/src/main.d.ts
@@ -1,28 +1,13 @@
-import type {
-  ErrorClass,
-  SpecificErrorClass,
-  ErrorSubclassCore,
-} from './subclass/create.js'
+import type { ErrorClass, SpecificErrorClass } from './subclass/create.js'
 import type { CustomClass } from './subclass/custom.js'
 
 export type { ErrorInstance } from './merge/cause.js'
 export type { ClassOptions } from './options/class.js'
 export type { InstanceOptions } from './options/instance.js'
 export type { MethodOptions } from './options/method.js'
-export type {
-  Info,
-  // Export required due to
-  // https://github.com/ehmicky/modern-errors/issues/18
-  CommonInfo,
-} from './plugins/info/main.js'
+export type { Info } from './plugins/info/main.js'
 export type { Plugin } from './plugins/shape/main.js'
-export type {
-  ErrorClass,
-  // Exports required due to
-  // https://github.com/ehmicky/modern-errors/issues/18
-  CustomClass,
-  ErrorSubclassCore,
-}
+export type { ErrorClass }
 
 /**
  * Top-level `ErrorClass`.
@@ -86,3 +71,11 @@ export default ModernError
 //    (e.g. `prepareStackTrace()`)
 //  - Plugins should not be allowed to define instance|static methods already
 //    defined by other plugins
+
+// NOTE: The following exports are temporary workarounds for:
+// - [modern-errors issue #18](https://github.com/ehmicky/modern-errors/issues/18)
+// - [TypeScript issue #47663](https://github.com/microsoft/TypeScript/issues/47663)
+export type { CommonInfo } from './plugins/info/main.js'
+export type { CustomClass } from './subclass/custom.js'
+export type { CreateSubclass, ErrorSubclassCore } from './subclass/create.js'
+export type { NormalizeError } from './subclass/normalize.js'

--- a/src/plugins/info/main.d.ts
+++ b/src/plugins/info/main.d.ts
@@ -3,6 +3,13 @@ import type { ErrorClass } from '../../subclass/create.js'
 
 /**
  * Properties shared by all `info` objects.
+ *
+ * @private This type is private and only exported as a temporary workaround
+ * for an open issue with TypeScript. It will be removed in a future release.
+ * See:
+ *
+ * - [modern-errors issue #18](https://github.com/ehmicky/modern-errors/issues/18)
+ * - [TypeScript issue #47663](https://github.com/microsoft/TypeScript/issues/47663)
  */
 interface CommonInfo<Options = never> {
   /**

--- a/src/subclass/create.d.ts
+++ b/src/subclass/create.d.ts
@@ -19,6 +19,13 @@ import type { NormalizeError } from './normalize.js'
 
 /**
  * `ErrorClass.subclass()`
+ *
+ * @private This type is private and only exported as a temporary workaround
+ * for an open issue with TypeScript. It will be removed in a future release.
+ * See:
+ *
+ * - [modern-errors issue #18](https://github.com/ehmicky/modern-errors/issues/18)
+ * - [TypeScript issue #47663](https://github.com/microsoft/TypeScript/issues/47663)
  */
 type CreateSubclass<
   PluginsArg extends Plugins,
@@ -44,6 +51,13 @@ type CreateSubclass<
 
 /**
  * Non-dynamic members of error classes
+ *
+ * @private This type is private and only exported as a temporary workaround
+ * for an open issue with TypeScript. It will be removed in a future release.
+ * See:
+ *
+ * - [modern-errors issue #18](https://github.com/ehmicky/modern-errors/issues/18)
+ * - [TypeScript issue #47663](https://github.com/microsoft/TypeScript/issues/47663)
  */
 interface ErrorSubclassCore<
   PluginsArg extends Plugins,

--- a/src/subclass/custom.d.ts
+++ b/src/subclass/custom.d.ts
@@ -9,6 +9,13 @@ import type { Plugins } from '../plugins/shape/main.js'
 
 /**
  * `custom` option
+ *
+ * @private This type is private and only exported as a temporary workaround
+ * for an open issue with TypeScript. It will be removed in a future release.
+ * See:
+ *
+ * - [modern-errors issue #18](https://github.com/ehmicky/modern-errors/issues/18)
+ * - [TypeScript issue #47663](https://github.com/microsoft/TypeScript/issues/47663)
  */
 export interface CustomClass {
   new (message: string, options?: InstanceOptions): Error

--- a/src/subclass/normalize.d.ts
+++ b/src/subclass/normalize.d.ts
@@ -12,6 +12,13 @@ import type { CustomClass } from './custom.js'
 
 /**
  * `ErrorClass.normalize()`.
+ *
+ * @private This type is private and only exported as a temporary workaround
+ * for an open issue with TypeScript. It will be removed in a future release.
+ * See:
+ *
+ * - [modern-errors issue #18](https://github.com/ehmicky/modern-errors/issues/18)
+ * - [TypeScript issue #47663](https://github.com/microsoft/TypeScript/issues/47663)
  */
 export type NormalizeError<
   PluginsArg extends Plugins,

